pkg/beacon/relay/groupselection/groupselection.go

```diff
@@ -56,223 +46,214 @@ func CandidateToNewGroup(
 	if err != nil {
 		return err
 	}
-	initialSubmissionTickets, reactiveSubmissionTickets, err :=
-		generateTickets(
-			newEntry.Bytes(),
-			staker.Address(),
-			availableStake,
-			chainConfig.MinimumStake,
-			naturalThreshold(chainConfig),
-		)
+
+	tickets, err := generateTickets(
+		newEntry.Bytes(),
+		staker.Address(),
+		availableStake,
+		chainConfig.MinimumStake,
+	)
 	if err != nil {
 		return err
 	}
 
-	logger.Infof(
-		"generated [%v] tickets for initial submission phase and [%v] "+
-			"tickets for reactive submission phase",
-		len(initialSubmissionTickets),
-		len(reactiveSubmissionTickets),
-	)
+	logger.Infof("generated [%v] tickets", len(tickets))
 
 	return startTicketSubmission(
-		initialSubmissionTickets,
-		reactiveSubmissionTickets,
+		tickets,
 		relayChain,
 		blockCounter,
 		chainConfig,
 		startBlockHeight,
 		onGroupSelected,
+		ticketSubmissionRoundDuration,
 	)
 }
 
 func startTicketSubmission(
-	initialSubmissionTickets []*ticket,
-	reactiveSubmissionTickets []*ticket,
+	tickets []*ticket,
 	relayChain relaychain.GroupSelectionInterface,
 	blockCounter chain.BlockCounter,
 	chainConfig *config.Chain,
 	startBlockHeight uint64,
 	onGroupSelected func(*Result),
+	ticketSubmissionRoundDuration uint64,
 ) error {
-	initialSubmissionTimeout, err := blockCounter.BlockHeightWaiter(
-		startBlockHeight + chainConfig.TicketSubmissionTimeout/2,
-	)
-	if err != nil {
-		return err
-	}
-
-	reactiveSubmissionTimeout, err := blockCounter.BlockHeightWaiter(
+	ticketSubmissionTimeoutChannel, err := blockCounter.BlockHeightWaiter(
 		startBlockHeight + chainConfig.TicketSubmissionTimeout,
 	)
 	if err != nil {
 		return err
 	}
 
-	// Buffer quit signals - we never know if the goroutine finished
-	// before we try to cancel it. The initial ticket submission may be
-	// cancelled right after the initial submission timeout and after the
-	// reactive submission timeout and it is possible it already completed.
-	// Hence, we buffer two quit signals. The reactive ticket submission
-	// is cancelled right after the reactive submission timeout. Here as well,
-	// we do not know if the goroutine already completed, so we need to buffer
-	// one quit signal.
-	quitInitialTicketSubmission := make(chan struct{}, 2)
-	quitReactiveTicketSubmission := make(chan struct{}, 1)
-
-	// Check how many tickets with values below the natural threshold has been
-	// generated and compare this number with the group size. Decide how many
-	// tickets should be submitted. It does not make sense to submit more
-	// tickets than the group size.
-	var numberOfTicketsToSubmit int
-	if len(initialSubmissionTickets) > chainConfig.GroupSize {
-		numberOfTicketsToSubmit = chainConfig.GroupSize
-	} else {
-		numberOfTicketsToSubmit = len(initialSubmissionTickets)
+	quitTicketSubmission := make(chan struct{})
+
+	ticketSubmissionRounds := (chainConfig.TicketSubmissionTimeout /
+		ticketSubmissionRoundDuration) - 2
```

Comment:
Why `-2`? Is it a mining lag? If so, let's make it a `const` next to `ticketSubmissionRoundDuration`.

Reply:
Also, isn't the number of rounds `(timeout - miningLag) / roundDuration`?

Reply:
> Why -2? Is it a mining lag? If so, let's make it a const next to ticketSubmissionRoundDuration.

Yeah this is a mining lag which consists of two rounds, I will extract it to a constant.

> Also, isn't the number of rounds (timeout - miningLag) / roundDuration?

This is basically the same.

Reply:
It would probably be better to calculate it as `(timeout - miningLag) / roundDuration` to cover the case where `roundDuration != miningLag / 2`